pkg/kubelet/images/pullmanager/fs_pullrecords.go

```diff
@@ -0,0 +1,275 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pullmanager
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/klog/v2"
+	kubeletconfigv1alpha1 "k8s.io/kubelet/config/v1alpha1"
+	kubeletconfiginternal "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	kubeletconfigvint1alpha1 "k8s.io/kubernetes/pkg/kubelet/apis/config/v1alpha1"
+)
+
+var _ PullRecordsAccessor = &fsPullRecordsAccessor{}
+
+// fsPullRecordsAccessor uses the filesystem to read/write ImagePullIntent/ImagePulledRecord
+// records.
+type fsPullRecordsAccessor struct {
+	pullingDir string
+	pulledDir  string
+
+	encoder runtime.Encoder
+	decoder runtime.Decoder
+}
+
+// NewFSPullRecordsAccessor returns an accessor for the ImagePullIntent/ImagePulledRecord
+// records with a filesystem as the backing database.
+func NewFSPullRecordsAccessor(kubeletDir string) (*fsPullRecordsAccessor, error) {
+	kubeletConfigEncoder, kubeletConfigDecoder, err := createKubeletConfigSchemeEncoderDecoder()
+	if err != nil {
+		return nil, err
+	}
+
+	accessor := &fsPullRecordsAccessor{
+		pullingDir: filepath.Join(kubeletDir, "image_manager", "pulling"),
+		pulledDir:  filepath.Join(kubeletDir, "image_manager", "pulled"),
+
+		encoder: kubeletConfigEncoder,
+		decoder: kubeletConfigDecoder,
+	}
+
+	if err := os.MkdirAll(accessor.pullingDir, 0700); err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(accessor.pulledDir, 0700); err != nil {
+		return nil, err
+	}
+
+	return accessor, nil
+}
+
+func (f *fsPullRecordsAccessor) WriteImagePullIntent(image string) error {
+	intent := kubeletconfiginternal.ImagePullIntent{
+		Image: image,
+	}
+
+	intentBytes := bytes.NewBuffer([]byte{})
+	if err := f.encoder.Encode(&intent, intentBytes); err != nil {
+		return err
+	}
+
+	return writeFile(f.pullingDir, cacheFilename(image), intentBytes.Bytes())
+}
+
+func (f *fsPullRecordsAccessor) ListImagePullIntents() ([]*kubeletconfiginternal.ImagePullIntent, error) {
+	var intents []*kubeletconfiginternal.ImagePullIntent
+	// walk the pulling directory for any pull intent records
+	err := processDirFiles(f.pullingDir,
+		func(filePath string, fileContent []byte) error {
+			intent, err := decodeIntent(f.decoder, fileContent)
+			if err != nil {
+				klog.V(4).InfoS("skipping file, failed to deserialize to ImagePullIntent", "filePath", filePath, "err", err)
+				return nil
```

Comment:
I *think* returning the error here is better than logging it, and processDirFiles will aggregate it (along with any read / walk errors)

Reply:
same comment applies everywhere we call processDirFiles